### PR TITLE
Document PHPStan usage in extensions

### PR DESCRIPTION
**Pull request by SychO9**
Goes hand-in-hand with: https://github.com/flarum/framework/pull/3666
This also needs updating the Flarum CLI before merging.

### CODE REVIEW

**Comment by n-peugnet** [comment]
This is very cool, can't wait to use it!

It seems another thing missing, besides the CLI changes, is to publish the package to Packagist repository. Could you do it, even without tagged version? I would love to try this out.

File: docs/extend/models.md
Comment (askvortsov1) [inline docs/extend/models.md:144]: Doesn't this comment need an end?

Comment (askvortsov1) [reply docs/extend/models.md:144]: Why not remove this section in favor of the new one?

File: docs/extend/static-code-analysis.md
Comment (askvortsov1) [inline docs/extend/static-code-analysis.md:55]: Shouldn't all of this be done by the CLI? If so, can we denote that you need to run the CLI command OR do this?

**Review by askvortsov1 — APPROVED** [review]
We can probably do that, and merge this, when we release 2.7

File: docs/extend/static-code-analysis.md
Comment (n-peugnet) [inline docs/extend/static-code-analysis.md:29]: This should likely be `flarum/phpstan` for extension developers:

```suggestion
  - vendor/flarum/phpstan/extension.neon
```

File: docs/extend/static-code-analysis.md
Comment (n-peugnet) [inline docs/extend/static-code-analysis.md:40]: This path also seems strange for out-of-tree extensions.

File: docs/extend/testing.md
Comment (n-peugnet) [inline docs/extend/testing.md:26]: Looks like this backtick should not stay here:

```suggestion
composer require --dev flarum/testing:^1.0
```

**Comment by n-peugnet** [comment]
I just remembered that I can simply install the package from the git repo while it is not released yet. So I tested it and it seems there is a problem with the `stubFiles` that are included from the `vendor` directory.

First I had to set the `include` path as so as I said above https://github.com/flarum/docs/pull/441#discussion_r1097712669:

```neon
# phpstan.neon
includes:
  - vendor/flarum/phpstan/extension.neon
```

But then I had the following error:

    Note: Using configuration file /home/nicolas/Source/www/flarum/packages/cross-references/phpstan.neon.
    Stub file /home/nicolas/Source/www/flarum/packages/cross-references/vendor/vendor/nunomaduro/larastan/stubs/Enumerable.stub does not exist.

And indeed, the paths of the stub files are wrong:

```neon
# vendor/flarum/phpstan/larastan-extension.neon
parameters:
    stubFiles:
        - ../../vendor/nunomaduro/larastan/stubs/Enumerable.stub
        - ../../vendor/nunomaduro/larastan/stubs/EloquentBuilder.stub
        - ../../vendor/nunomaduro/larastan/stubs/Collection.stub
        - ../../vendor/nunomaduro/larastan/stubs/EloquentCollection.stub

        [...]
```

This is because the way its used internally in Flarum source tree is one less level deep than what it will be once installed through composer.

One easy way I see to fix this without having to move the directory of the extension in `flarum/framwork` is to use the [`%rootDir%` path expansion from PHPStan](https://phpstan.org/config-reference#expanding-paths) (for instance `%rootDir%/../../vendor/nunomaduro/larastan/stubs/Enumerable.stub`).

**Comment by n-peugnet** [comment]
I also have the following error messages:

```
Ignored error pattern #Relation '[A-z_-]+' is not found in [A-z ]+ model.# was not matched in reported errors.                                            
Ignored error pattern #^Parameter \#1 \$query of method [A-z_<>\\]+\:\:union\(\) expects [A-z_<> .,|\\]+ given\.$# was not matched in reported errors.    
Ignored error pattern #^Parameter \#1 \$query of method [A-z_<>\\]+\:\:joinSub\(\) expects [A-z_<> .,|\\]+ given\.$# was not matched in reported errors.  
Ignored error pattern #Template type T of function resolve[()]{2} is not referenced in a parameter.# was not matched in reported errors.                  
Ignored error pattern #^Call to an undefined method Illuminate\\Database\\ConnectionInterface\:\:[A-z0-9_]+\(\)\.$# was not matched in reported errors.   
```

This comes from the `phpstan-baseline.neon`:

```neon
# vendor/flarum/phpstan/phpstan-baseline.neon
parameters:
	ignoreErrors:
		# Remove this group below with larastan 2.0 (i.e Flarum 2.0)
		- "#Relation '[A-z_-]+' is not found in [A-z\_]+ model.#"
		- '#^Parameter \#1 \$query of method [A-z_<>\\]+\:\:union\(\) expects [A-z_<> .,|\\]+ given\.$#'
		- '#^Parameter \#1 \$query of method [A-z_<>\\]+\:\:joinSub\(\) expects [A-z_<> .,|\\]+ given\.$#'

		#	We ignore this because resolve can either take a class name as the generic return type or just a binding name.
		- "#Template type T of function resolve[()]{2} is not referenced in a parameter.#"

		#	We ignore new static errors because we want extensibility.
		#	@TODO: needs discussion.
		- "#^Unsafe usage of new static[()]{2}.$#"

		# ConnectionInterface lacks methods that exist in the implementation,
		# yet we don't want to inject the implementation.
		- '#^Call to an undefined method Illuminate\\Database\\ConnectionInterface\:\:[A-z0-9_]+\(\)\.$#'

```

It might be a good thing to [selectively disable reporting unused ignore](https://phpstan.org/user-guide/ignoring-errors#reporting-unused-ignores) for each of these ignoreErrors like this:

```neon
parameters:
	ignoreErrors:
		# Remove this group below with larastan 2.0 (i.e Flarum 2.0)
		- message: "#Relation '[A-z_-]+' is not found in [A-z\_]+ model.#"
		  reportUnmatched: false

		[...]
```

File: docs/extend/models.md
Comment (SychO9) [reply docs/extend/models.md:144]: cause it references a non existent `accessor` and `mutator` extenders that kind of act as a reminder to actually add them 😋

File: docs/extend/static-code-analysis.md
Comment (SychO9) [reply docs/extend/static-code-analysis.md:55]: When the CLI gets augmented to actually do this we will update this doc page ;)

File: docs/extend/static-code-analysis.md
Comment (SychO9) [reply docs/extend/static-code-analysis.md:40]: good catch!
```suggestion
	databaseMigrationsPath: ['migrations']
```

File: docs/extend/static-code-analysis.md
Comment (SychO9) [reply docs/extend/static-code-analysis.md:55]: actually, I already added a future command reference on the top of the page

**Review by n-peugnet — APPROVED** [review]

File: docs/extend/static-code-analysis.md
Comment (n-peugnet) [inline docs/extend/static-code-analysis.md:86]: Just one question about this section. The last error that I couldn't fix is:

    Call to an undefined method Flarum\Discussion\Discussion::sources()

This is a [method I added in extend.php](https://github.com/club-1/flarum-ext-cross-references/blob/2731136be37dcdbabbd41846b6007d2d097a2a25/extend.php#L48-L50) like this:

```php
    (new Extend\Model(Discussion::class))
        ->belongsToMany('sources', Discussion::class, 'discussion_reference', 'target_id', 'source_id')
        ->belongsToMany('targets', Discussion::class, 'discussion_reference', 'source_id', 'target_id'),
```

How can I make PHPStan understand method declared like this?

I thought about using the `cast()` method as it seems it is supposed to solve this kind of issues but it looks like is is only for primitive types.

File: docs/extend/static-code-analysis.md
Comment (n-peugnet) [inline docs/extend/static-code-analysis.md:40]: Always use 2 space to keep indentation consistent in the example neon file:

```suggestion
includes:
  - vendor/flarum/phpstan/extension.neon

parameters:
  # The level will be increased in Flarum 2.0
  level: 5
  paths:
    - src
    - extend.php
  excludePaths:
    - *.blade.php
  checkMissingIterableValueType: false
  databaseMigrationsPath: ['migrations']
```

File: docs/extend/static-code-analysis.md
Comment (SychO9) [reply docs/extend/static-code-analysis.md:86]: The package should have logic for all relationship methods, I need to check again `belongsToMany` might have been fogotten

Comment (SychO9) [reply docs/extend/static-code-analysis.md:86]: @n-peugnet can you link to the codebase, `belongsToMany` is correctly picked up by the phpstan package when used on the framework, so not sure why it isn't working on your end.

Comment (n-peugnet) [reply docs/extend/static-code-analysis.md:86]: The extension is this one : https://github.com/club-1/flarum-ext-cross-references

I added phpstan setup in this branch : https://github.com/club-1/flarum-ext-cross-references/pull/25

For now it fails with this issue : https://github.com/club-1/flarum-ext-cross-references/actions/runs/4143686829/jobs/7165830767#step:4:15

    Stub file /home/runner/work/flarum-ext-cross-references/flarum-ext-cross-references/vendor/vendor/nunomaduro/larastan/stubs/Enumerable.stub does not exist.

As my commit is not yet available in [flarum/phpstan](https://github.com/flarum/phpstan).

Comment (SychO9) [reply docs/extend/static-code-analysis.md:86]: @n-peugnet it is now

Comment (n-peugnet) [reply docs/extend/static-code-analysis.md:86]: Thanks, I will push a new commit right now.

Comment (n-peugnet) [reply docs/extend/static-code-analysis.md:86]: @SychO9 I just updated `flarum/phpstan` to the latest version and now everything is working fine. The stub files are now found as expected and mysteriously, the `Call to an undefined method Flarum\Discussion\Discussion::sources()` error disappeared.

Well, good for us I guess!

Comment (SychO9) [reply docs/extend/static-code-analysis.md:86]: nice :) thanks for the help!

**Pull request merged**
